vaadin-grid-flow-parent/vaadin-grid-flow/src/main/java/com/vaadin/flow/component/grid/AbstractGridMultiSelectionModel.java

```diff
@@ -437,12 +446,13 @@ private void doUpdateSelection(Map<Object, T> addedItems,
 
         fireSelectionEvent(new MultiSelectionEvent<>(getGrid(),
                 getGrid().asMultiSelect(), oldSelection, userOriginated));
-        if (!removedItems.isEmpty()) {
-            selectionColumn.setSelectAllCheckboxState(false);
-        } else {
-            long size = getDataProviderSize();
-            selectionColumn.setSelectAllCheckboxState(size == selected.size());
-        }
+
+        long size = getDataProviderSize();
```

Comment:
Same here, the previous optimization of not calling the data provider when at least one item is removed does not suffice anymore when trying to calculate the indeterminate state. When refactoring we could introduce other optimizations instead, for example pass in additional information into this method, whether a select all / deselect triggered this update.